networkx/algorithms/bidirectional_path_search.py

```diff
@@ -0,0 +1,505 @@
+                             /\                                        |
+                            /\ \                                       |
+                           / /\ \                                      |
+                          /   /\ \                                     |
+                         /\  /\  /\                                    |
+                        /  \/   /  \                                   |
+                       /    \  /\   \                                  |
+                      /_____/_/__\___\_________   meeting points       |
+                     /\ \   \    /   /\                                |
+                    /XX\ \  /  \/   /XX\                               |
+                   /XXXX\ \ \/  \  /XXXX\                              |
+                  /XXXXXX\ \/    \/XXXXXX\                             |
+                 /XXXXXXXX\ \/   /XXXXXXXX\                            |
+                /XXXXXXXXXX\ \  /XXXXXXXXXX\                           |
+               /XXXXXXXXXXXX\ \/XXXXXXXXXXXX\                          |
+              /XXXXXXXXXXXXXX\/XXXXXXXXXXXXXX\                         |
+                           target                                      |
+    Authors
+    -------
+    André Dietrich and Sebastian Zug
+    {dietrich, zug}@ivs.cs.uni-magdeburg.de
+
+    See Also
+    --------
+    nx.all_simple_paths, nx.all_shortest_paths, nx.shortest_path,
+    nx.bidirectional_all_shortest_paths, nx.bidirectional_has_path
+    """
+
+    if source not in G:
+        raise nx.NetworkXError('source node %s not in graph' % source)
+    if target not in G:
+        raise nx.NetworkXError('target node %s not in graph' % target)
+    if cutoff is None:
+        cutoff = len(G) - 1
+    if G.is_multigraph():
+        return _bidirectional_all_simple_paths_multi(G, source, target, cutoff=cutoff)
+    else:
+        return _bidirectional_all_simple_paths(G, source, target, cutoff=cutoff)
+
+
+def _bidirectional_all_simple_paths(G, source, target, cutoff):
+
+    if cutoff < 1:
+        return
+
+    tree = [{(source,)}, {(target,)}]
+    leaves = [{source}, {target}]
+
+    # used if the graph is not directed
+    neighbors = lambda node: iter(G[node])
+
+    for i in range(1, cutoff+1):
+
+        if len(tree[1]) < len(tree[0]):
+            modus = 1
+            tree_source = tree[1]
+            tree_target = tree[0]
+            leaves_ = leaves[0]
+
+            if G.is_directed():
+                neighbors = G.predecessors
+        else:
+            modus = 0
+            tree_source = tree[0]
+            tree_target = tree[1]
+            leaves_ = leaves[1]
+
+            if G.is_directed():
+                neighbors = G.successors
+
+        temp_tree = set()
+        temp_leaves = set()
+
+        for path_s in iter(tree_source):
+            for s in neighbors(path_s[-1]):
+                if s not in path_s:
```

Comment:
The level of nesting is still too large to be readable. Perhaps at this point, do

    if s in path_s:
        continue
    if s in leaves_:
        for path_t in tree_target:
            if s != path_t[-1]:
                continue
            if ...